substrate/frame/revive/fixtures/contracts/Terminate.sol

```diff
@@ -4,40 +4,45 @@ pragma solidity ^0.8.30;
 import "@revive/ISystem.sol";
 
 contract Terminate {
-	constructor(bool skip) {
+	uint8 public constant METHOD_PRECOMPILE = 0;
+	uint8 public constant METHOD_DELEGATE_CALL = 1;
+	uint8 public constant METHOD_SYSCALL = 2;
+    receive() external payable {}
+	constructor(bool skip, uint8 method, address beneficiary) payable {
 		if (skip) {
 			return;
 		}
-		_terminate(false);
+		_terminate(method, beneficiary);
 	}
 
-	function terminate() external {
-		_terminate(false);
+	function terminate(uint8 method, address beneficiary) external {
+		_terminate(method, beneficiary);
 	}
 
-	function delegateTerminate() external {
-		_terminate(true);
-	}
-
-	function indirectDelegateTerminate() external {
-		bytes memory data = abi.encodeWithSelector(this.terminate.selector);
+	function indirectDelegateTerminate(address beneficiary) external {
+		bytes memory data = abi.encodeWithSelector(this.terminate.selector, METHOD_PRECOMPILE, beneficiary);
 		(bool success, bytes memory returnData) = address(this).delegatecall(data);
 		if (!success) {
 			assembly {
 				revert(add(returnData, 0x20), mload(returnData))
 			}
 		}
 	}
-
 	// Call terminate and forward any revert
-	function _terminate(bool delegate) private {
-		bytes memory data = abi.encodeWithSelector(ISystem.terminate.selector, address(0));
+	function _terminate(uint8 method, address beneficiary) private {
+		bytes memory data = abi.encodeWithSelector(ISystem.terminate.selector, beneficiary);
 		(bool success, bytes memory returnData) = (false, "");
 
-		if (delegate) {
+		if (method == METHOD_DELEGATE_CALL) {
 			(success, returnData) = SYSTEM_ADDR.delegatecall(data);
-		} else {
+		} else if (method == METHOD_PRECOMPILE) {
 			(success, returnData) = SYSTEM_ADDR.call(data);
+		} else if (method == METHOD_SYSCALL) {
+			assembly {
+				selfdestruct(beneficiary)
+			}
```

Comment:
I think this could use an explicit revert after the assembly block: We are not expected to ever hit this. But if execution somehow continues despite the selfdestruct it would be bad and we would catch that here.